addons/binding/org.openhab.binding.folderwatcher/src/main/java/org/openhab/binding/folderwatcher/internal/handler/FolderLocalWatcherHandler.java

```diff
@@ -0,0 +1,234 @@
+        }
+
+    }
+
+    @Override
+    public void dispose() {
+        executionJob.cancel(true);
+
+    }
+    /*
+     * private void listNewFiles() throws IOException {
+     *
+     * ArrayList<String> diffFtpListing = new ArrayList<String>();
+     *
+     * if (!currentLocalListing.isEmpty()) {
+     * FileWriter fileWriter = new FileWriter(currentLocalListingFile);
+     *
+     * for (String newFtpFile : currentLocalListing) {
+     * fileWriter.write(newFtpFile + "\n");
+     * }
+     *
+     * fileWriter.close();
+     *
+     * diffFtpListing = (ArrayList<String>) currentLocalListing.clone();
+     * diffFtpListing.removeAll(previousLocalListing);
+     *
+     * for (String newFtpFile : diffFtpListing) {
+     *
+     * triggerChannel(CHANNEL_LOCALFILENAME, newFtpFile);
+     *
+     * try {
+     * Thread.sleep(3000);
+     * } catch (InterruptedException e) {
+     *
+     * }
+     * }
+     * previousLocalListing = (ArrayList<String>) currentLocalListing.clone();
+     * }
+     * }
+     */
+
+    void listFiles(String path, boolean listHidden, boolean recursive) throws IOException {
+        File f = null;
+        File[] fileList;
+
+        try {
+
+            // create new file
+            f = new File(path);
+
+            // array of files and directory
+
+            fileList = f.listFiles(new FileFilter() {
+                @Override
+                public boolean accept(@Nullable File fileRaw) {
+                    if (listHidden) {
+                        return true;
+                    } else {
+                        return !fileRaw.isHidden();
+                    }
+                }
+            });
+
+            // for each name in the path array
+            if (fileList != null && fileList.length > 0) {
+                for (File file : fileList) {
+
+                    if (file.isDirectory()) {
+
+                        if (recursive) {
+                            listFiles(file.getAbsolutePath(), listHidden, recursive);
+                        }
+                    }
+                    if (file.isFile()) {
+                        // logger.debug(file.getAbsolutePath());
+                        currentLocalListing.add(file.getAbsolutePath());
+                    }
+                }
+            }
+
+        } catch (Exception e) {
+            throw e;
+        }
+    }
+
+    protected Runnable periodicExecutionRunnable = new Runnable() {
+        @Override
+        public void run() {
+
+            String RootDir = config.localDir;
+            ArrayList<String> diffLocalListing = new ArrayList<String>();
+
+            /*
+             * if (ftpRootDir.endsWith("/")) {
+             * ftpRootDir = ftpRootDir.substring(0, ftpRootDir.length() - 1);
+             * }
+             *
+             * if (!ftpRootDir.startsWith("/")) {
+             * ftpRootDir = "/" + ftpRootDir;
+             * }
+             */
+            currentLocalListing.clear();
```

Comment:
I think it would be better to first get the new listing and when everything is done overwrite the complete listing.
Otherwise the listing could end up in an inconsistent state in case of errors

Reply:
Fixed